### PR TITLE
feat(sidenav): add responsive sidenav

### LINKED ISSUE #1130: md-sidenav: make it automatically responsive
*Feature request*
Version: Angular Material 2.0.0-alpha.7 wax-umpire

I'd like to have an option where the sidenav can be responsive automatically, by having a breakpoint where the sidenav is shown by default when the width is higher than X. 

This was already implemented in Angular Material (1) as the `md-is-locked-open`: https://material.angularjs.org/latest/api/directive/mdSidenav

Is there any plan to have this? I could submit a PR if this is wanted.

### PR DESCRIPTION
**Pull request by mcparadip**
* Add responsive sidenav
* Close and open depending on width

Fix #1130

### CODE REVIEW

File: src/lib/sidenav/sidenav.ts [hunk 1]
```diff
@@ -316,6 +317,35 @@ export class MdSidenavContainer implements AfterContentInit {
...
+
+  /** Mode of the breakpoint; either true or false */
+  @Input() breakpointChangeMode: true | false = true;
+  
+  /**
+   * Responsively toggle the sidenav using the breakpoint.
+   * Note that this only works when the window is rezied.
+   * If you resize it some other way, call checkBreakpoint().
+   */
+  @HostListener('window:resize', ['$event'])
```
Comment (willshowell) [inline src/lib/sidenav/sidenav.ts]: This should be debounced. See https://github.com/angular/material2/blob/master/src/lib/tabs/tab-header.ts#L182-L195

File: src/lib/sidenav/sidenav.ts [hunk 2]
```diff
@@ -316,6 +317,35 @@ export class MdSidenavContainer implements AfterContentInit {
...
   get end() { return this._end; }
 
+  /**
+   * Width of the container at which it breaks
+   * e.g., breakpointWidth="500"
+   */
+  @Input() breakpointWidth: number;
+
+  /** Mode of the breakpoint; either true or false */
+  @Input() breakpointChangeMode: true | false = true;
```
Comment (willshowell) [inline src/lib/sidenav/sidenav.ts]: Should be a `boolean` and use `coerceBooleanProperty`

Comment (willshowell) [inline src/lib/sidenav/sidenav.ts]: Also why not just look at whether the `breakpointWidth` is defined?

Comment (mcparadip) [inline src/lib/sidenav/sidenav.ts]: The mode is if you want to change the mode to "over" when it is collapsed by default. Will fix boolean thing

Comment (willshowell) [inline src/lib/sidenav/sidenav.ts]: Perhaps update the comment to be more like what you just told me. It isn't clear from "Mode of the breakpoint"

Comment (mcparadip) [inline src/lib/sidenav/sidenav.ts]: This good?


      /**
       * Mode of the breakpoint; either true or false
       * If true, the mode of the sidenav will change to "over" when it
       * is collapsed by default, and "side" when it is not.
       */

Comment (willshowell) [inline src/lib/sidenav/sidenav.ts]: Try this:

```ts
/** Whether the the drawer changes modes when collapsing. */
```

Note: the terminology changes from sidenav to drawer in a pretty recent commit

File: src/lib/sidenav/sidenav.ts [hunk 3]
```diff
@@ -316,6 +317,35 @@ export class MdSidenavContainer implements AfterContentInit {
...
+  /**
+   * Responsively toggle the sidenav using the breakpoint.
+   * Note that this only works when the window is rezied.
+   * If you resize it some other way, call checkBreakpoint().
+   */
+  @HostListener('window:resize', ['$event'])
+  checkBreakpoint() {
+    if (this._element.nativeElement.offsetWidth < this.breakpointWidth) {
+      if (this.breakpointChangeMode)
+        this._sidenavs.forEach(sidenav => sidenav.mode = "over");
```
Comment (willshowell) [inline src/lib/sidenav/sidenav.ts]: Use curly brackets here and below

File: src/lib/sidenav/sidenav.ts (hunk 2 above)
Comment (mcparadip) [inline src/lib/sidenav/sidenav.ts]: ok :/

I have to change all my `sidenav`s to `drawer`s now? I looked in that commit and there still seems to be some kind of sidenav, what is the difference?

Comment (mcparadip) [inline src/lib/sidenav/sidenav.ts]: I'm going to start a new fork because the difference is big and I don't like pulling

**Comment by mcparadip** [comment]
Starting new branch because of drawer/sidenav being split.

**Comment by mcparadip** [comment]
New version: #6535

**Pull request closed without merging**